scour/scour.py

```diff
@@ -3239,12 +3229,24 @@ def serializeXML(element, options, ind=0, preserveWhitespace=False):
     attrIndices += [attrName2Index[name] for name in sorted(attrName2Index.keys())]
     for index in attrIndices:
         attr = attrList.item(index)
-        # if the attribute value contains a double-quote, use single-quotes
-        quot = '"'
-        if attr.nodeValue.find('"') != -1:
-            quot = "'"
 
-        attrValue = makeWellFormed(attr.nodeValue)
+        attrValue = attr.nodeValue
+
+        quot_count = 0
+        apos_count = 0
+
+        for c in attrValue:
+            if c == '"':
+                quot_count += 1
+            elif c == "'":
+                apos_count += 1
+
```

Comment:
Any reason for using a for-loop here?

Can't we simply use [str.count](https://docs.python.org/3/library/stdtypes.html#str.count)?
(If it's significantly faster I'm fine with it, but I don't want to decrease legibility if it's only some negligible speedup).

Reply:
Interestingly, my initial code did indeed use 2 calls to `str.count`, but as a matter of principle, I just couldn't stand the thought of looping through the same data twice (serially, no less!), when one loop would be enough.

Oh! Were that Python a more expressive language, with a more sophisticated interpreter! It'd be glorious to specify 2 calls to `str.count`, knowing that some kind of JIT would determine at run-time either to collapse them into one loop or spin them off into separate, parallel threads. Alas, our tools are not yet so beautiful.

Thus, I advocate for the single loop; however, how about abstracting that loop out into a function:

    def countQuotes(str):
        quotCount = 0
        aposCount = 0
    
        for c in str:
            if c == '"':
                quotCount += 1
            elif c == "'":
                aposCount += 1
    
        return (quotCount, aposCount)
    
    
    def chooseQuoteCharacter(str):
        (quotCount, aposCount) = countQuotes(str)
        quote = "'" if quotCount > aposCount else '"'
        return (quote, quotCount or aposCount)

Then, the code becomes something like:

    attrValue = attr.nodeValue
    (quote, hasEmbeddedQuote) = chooseQuoteCharacter(attrValue)
    attrValue = makeWellFormed(attrValue, quote if hasEmbeddedQuote else '')

Reply:
Let's do it with `count`... Unless I measured something wrong it's a lot faster:

```
import timeit

setup = '''
attrValue = '\'\"Where\'s www.wikipedia.org?\"\''
'''

a = '''
quot_count = 0
apos_count = 0

for c in attrValue:
    if c == '"':
        quot_count += 1
    elif c == "'":
        apos_count += 1

quote = "'" if quot_count > apos_count else '"'
'''

b = '''
quot_count = 0
apos_count = 0

quot_count = attrValue.count("'")
apos_count = attrValue.count('"')

quote = "'" if quot_count > apos_count else '"'
'''

number = 1000000
print(timeit.timeit(stmt=a, setup=setup, number=number))
print(timeit.timeit(stmt=b, setup=setup, number=number))
```

Reply:
Well, OK. I can't argue with those numbers.